src/relay/op/nn/upsampling.cc

```diff
@@ -18,6 +18,31 @@ namespace relay {
 
 TVM_REGISTER_NODE_TYPE(UpSamplingAttrs);
 
+template <typename T>
+Array<Array<Layout> > UpsamplingInferCorrectLayout(
+    const Attrs& attrs,
+    const Array<Layout>& new_in_layouts,
+    const Array<Layout>& old_in_layouts,
+    const Array<Array<IndexExpr>> &old_in_shapes) {
+  // NOTE: Discard "const" qualifier here.
+  T *params = const_cast<T*>(attrs.as<T>());
```

Comment:
Could you explain why this is necessary? It's quite counterintuitive why this kind of thing is needed, since the desired behaviour is essentially "if input layout is NCHWc, then expect input in NCHWc, perform compute, and output in NCHWc" right?

Reply:
This function is same to: https://github.com/dmlc/tvm/blob/master/src/relay/op/nn/pooling.cc#L22

In order to invoke AlterLayout Pass, we need ```FInferCorrectLayout``` attr, which is this function. This is part of logic of current AlterLayoutPass.